src/geom/SkyWcs.cc

```diff
@@ -0,0 +1,249 @@
+    auto oldGridToPixel0Map = frameSet->findFrame(ast::Frame(2, "Domain=PIXEL0"));
+    if (!oldGridToPixel0Map) {
+        throw LSST_EXCEPT(lsst::pex::exceptions::RuntimeError, "PIXEL0 frame not found");
+    }
+    auto oldPixel0Index = frameSet->getCurrent();
+    auto pixel0Frame = frameSet->getFrame(oldPixel0Index);
+
+    // Look for the optional ACTUAL_PIXEL0 frame, and the mapping from PIXEL0 to ACTUAL_PIXEL0
+    frameSet->setBase(oldPixel0Index);
+    auto oldActualPixel0Index = ast::FrameSet::NOFRAME;
+    std::shared_ptr<ast::Frame> actualPixel0Frame;
+    auto pixelToActualPixelMap = frameSet->findFrame(ast::Frame(2, "Domain=ACTUAL_PIXEL0"));
+    if (pixelToActualPixelMap) {
+        actualPixel0Frame = frameSet->getFrame(ast::FrameSet::CURRENT);
+        oldActualPixel0Index = frameSet->getCurrent();
+    }
+
+    // Compute new mapping from GRID to PIXEL0
+    std::vector<double> shiftVec = {dx, dy};
+    auto deltaShiftMap = ast::ShiftMap(shiftVec);
+    auto newGridToPixel0Map = deltaShiftMap.of(*oldGridToPixel0Map).simplify();
+
+    // Replace the mapping from GRID to PIXEL0.
+    // We actually remove the old PIXEL0 and ACTUAL_PIXEL0 frames (if present)
+    // and then add them in again, since there no direct way to replace a mapping.
+    frameSet->removeFrame(oldPixel0Index);
+    if (gridFrameIndex > oldPixel0Index) {
+        --gridFrameIndex;
+    }
+    if (skyFrameIndex > oldPixel0Index) {
+        --skyFrameIndex;
+    }
+    if (pixelToActualPixelMap) {
+        if (oldActualPixel0Index > oldPixel0Index) {
+            --oldActualPixel0Index;
+        }
+        frameSet->removeFrame(oldActualPixel0Index);
+        if (gridFrameIndex > oldActualPixel0Index) {
+            --gridFrameIndex;
+        }
+        if (skyFrameIndex > oldActualPixel0Index) {
+            --skyFrameIndex;
+        }
+    }
+    frameSet->addFrame(gridFrameIndex, *newGridToPixel0Map, *pixel0Frame);
+
+    // If ACTUAL_PIXEL0 exists, re-add it
+    if (pixelToActualPixelMap) {
+        frameSet->addFrame(ast::FrameSet::CURRENT, *pixelToActualPixelMap, *actualPixel0Frame);
+    }
+
+    // Set the last frame added (PIXEL0 or ACTUAL_PIXEL0) to the base frame
+    // and the SkyFrame to the current frame
+    frameSet->setBase(frameSet->getCurrent());
+    frameSet->setCurrent(skyFrameIndex);
+    return SkyWcs(std::move(frameSet));
+}
+
+std::pair<Angle, Angle> SkyWcs::pixelToSky(double x, double y) const {
+    auto sky = tranForward(Point2D(x, y));
+    return std::pair<Angle, Angle>(sky[0], sky[1]);
+}
+
+/**
+Compute the pixel position from the sky position
+*/
+std::pair<double, double> SkyWcs::skyToPixel(Angle const& ra, Angle const& dec) const {
+    auto pixel = tranInverse(SpherePoint(ra, dec));
+    return std::pair<double, double>(pixel[0], pixel[1]);
+};
+
+SkyWcs::SkyWcs(std::shared_ptr<ast::FrameSet> &&frameSet)
+        : Transform<Point2Endpoint, SpherePointEndpoint>(std::move(frameSet)){};
```

Comment:
clang-format. As to complexity, I consider it messy enough that I'd rather leave it out of the header.